### ISSUE TITLE
Namespaced Nodes

### ISSUE THREAD

**Issue by lordsoffallen** [kedro-org/kedro#3679]
## Description

I was under the impression that when I namespace a node, then I was able to run it via `kedro run --nodes mynode --namespace ns`. However kedro always looks for `ns.mynode` so passing just the node name is not enough. Shouldn't the behavior  be other way around? Me as a user, I don't have to know about kedro's special syntax for namespacing so passing that should take care of the rest, no?

Please correct the issue type if it's not considered a bug.

## Actual Result

```
-- If you received an error, place it here.
```

```
-- Separate them if you have more than one.
```

## Your Environment

* Kedro version used (`pip show kedro` or `kedro -V`):
* Python version used (`python -V`):
* Operating system and version:

**Comment by lordsoffallen** [comment]
It would be nice to support this case as this would reduce some duplication. I use namespace to connect different data sources to same functions. I can give them different names but it looks way better via namespacing in terms of name consistency

**Comment by noklam** [comment]
@datajoely is correct that it's more like AND, there are more use cases that are only possible with Pipeline API but not CLI, [this is one of the example](https://noklam.github.io/blog/posts/kedro-pipeline-slicing-pipeline/2024-03-06-Kedro-Pipeline-Slicing-Pipeline.html#advance-pipeline-slicing). On the other hand, seems like very little people structure pipelines in a way that requires complex filtering (I asked about this in Slack but very few responses). From my observation people usually use one of the filter, rarely using many at the same time. i.e. `kedro run -p <pipeline>` or `kedro run --tags <tag1,tag2>` 

Can you elaborate on how your pipelines/nodes are structured? I cannot wrap it in my head yet. I think there are value to enable this but it shouldn't disrupt the existing workflow as it seems to work for most people.

**Comment by lordsoffallen** [comment]
When I think about the namespace, I remember k8s generally. Basically same nodes/pipelines can be stored under different namespace. This helps running flow where user can use the same function but input/output different datasets.

Imagine you have a function which takes and filters some data on a column. Same column exists in other ones so you wanna use the same function for another dataset this time. I can def do it by giving nodes a different name like:

```
node(myfunc, input='data1', output='out1', name='myfunc1')
node(myfunc, input='data2', output='out2', name='myfunc2')
```

It's just that I figured namespace would provide this instead of me doing a different naming. So I expected this instead:

```
node(myfunc, input='data1', output='out1', name='myfunc', namespace='1')
node(myfunc, input='data2', output='out2', name='myfunc', namespace='2')
```
This way I could just filter other nodes by using the namespace. 

Regarding the -p pipeline filter, I mainly use that to group certain nodes together. For instance, data processing nodes are moved to data pipeline and model logic is in the model pipeline which helps me run all with a single argument.

**Comment by noklam** [comment]
@lordsoffallen I think I understand where is this coming from. I am not sure if this is clearly better.

```
node(myfunc, input='data1', output='out1', name='myfunc', namespace='a')
node(myfunc, input='data2', output='out2', name='other_function', namespace='b')
```

How should the CLI work in an unambiguous way?
```
kedro run --nodes a.myfunc,b.other_function
kedro run --nodes ??? --namespace a,b
```

One may argue it's clearly bad to apply multiple namespace here. But then it's perfectly reasonable to select multiple namespace https://github.com/kedro-org/kedro/issues/3056
```
kedro run --namespace a,b (This is actually not supported yet, I don't know why)
```

**Comment by lordsoffallen** [comment]
@noklam Your example involves different function names, that case wouldnt require a new namespace imo but if a user did that, following should work

```
kedro run --nodes myfunc --namespace a
kedro run --nodes other_func --namespace b
OR run both
kedro run --namespace a,b
```

For me namespacing should be similar to how it is in k8s if that makes sense. 

Ideally same function with different input/output would be super useful to support.

**Comment by astrojuanlu** [comment]
What was our most recent stance on namespacing nodes @DimedS ?

**Comment by DimedS** [comment]
Currently, when you assign a `namespace` to a pipeline or node in Kedro, all namespaced nodes are automatically renamed. Their `name` attribute gets updated to something like:

```
namespace_name.original_node_name
```

I agree that this can be unintuitive, especially when you define a node in code like:

```python
node(func, inputs, outputs, name="node_name1", namespace="namespace_name1")
```

You might expect the node's name to remain `node_name1`, but Kedro will change it to `namespace_name1.node_name1`.

As I understand it, @lordsoffallen would like to retain the original name alongside the namespace. However, this is currently not possible in Kedro - once the name is updated with the namespace prefix, the original name is no longer stored or accessible.

I believe node names are a fundamental concept in Kedro. They must remain unique to ensure consistency, especially across deployments and tools. There is an ongoing discussion that touches on these design considerations:  
https://github.com/kedro-org/kedro/pull/4715

To me, this is both a technical and design question. From a UI perspective, it would sometimes be helpful to treat `name` and `namespace` as separate fields rather than combining them. However, this separation is not currently supported, and it's also unclear how it *should* behave - especially since namespaces can be nested, making it harder to define what the "original" name truly is.

Curious to hear your thoughts, @iamelijahko.

**Comment by lordsoffallen** [comment]
Yes, i think the confusion was coming from there. I thought i could simply do `kedro run --nodes myfunc --namespace a` and it would run the node named myfunc in a specific namespace. Like namespace being a folder structure and nodes are the filenames per se.

Cuz right now, i have catalog entries that are generic and they follow a `prefix.filename` pattern so i don't know how the output will even look like. I sort of thought it would be a kedro internal thing but this namespace addition gets spilled onto everything.

Right now i am wrapping things into different pipelines so i can still have the same func name.
